rad/src/lib.rs

```diff
@@ -201,11 +208,81 @@ pub async fn run_retrieval_report(
 /// Run retrieval stage of a data request, return `RadonTypes`.
 pub async fn run_retrieval(retrieve: &RADRetrieve) -> Result<RadonTypes> {
     // Disable all execution tracing features, as this is the best-effort version of this method
-    run_retrieval_report(retrieve, RadonScriptExecutionSettings::disable_all())
+    run_retrieval_report(retrieve, RadonScriptExecutionSettings::disable_all(), None)
         .await
         .map(RadonReport::into_inner)
 }
 
+/// Run retrieval using multiple transports, and only produce a positive result if the retrieved
+/// values pass the filter function from the tally stage.
+///
+/// The idea behind this is to avoid producing commitments for data requests with sources that act
+/// in an inconsistent way, i.e. they return very different values when queried through different
+/// HTTP transports at once.
+pub async fn run_paranoid_retrieval(
+    retrieve: &RADRetrieve,
+    tally: &RADTally,
+    transports: &[Option<String>],
+    settings: RadonScriptExecutionSettings,
+) -> Result<RadonTypes> {
+    let futures: Vec<_> = transports
+        .iter()
+        .map(|transport| {
+            let client = WitnetHttpClient::new(transport).as_surf_client();
+
+            run_retrieval_report(
+                retrieve,
+                RadonScriptExecutionSettings::disable_all(),
+                Some(client),
+            )
+        })
+        .collect();
+
+    let values = join_all(futures).await;
+
+    evaluate_paranoid_retrieval(values, tally, settings)
+}
+
+/// Evaluate whether the values obtained when retrieving a data source through multiple transports
+/// are consistent, i.e. they all pass the filters from the tally stage.
+fn evaluate_paranoid_retrieval(
+    data: Vec<Result<RadonReport<RadonTypes>>>,
+    tally: &RADTally,
+    settings: RadonScriptExecutionSettings,
+) -> Result<RadonTypes> {
+    let mut values = vec![];
+
+    // Turn the reports into values, short-circuiting if any retrieval inevitably failed.
+    for each in data {
+        let report = each?;
+        let result = report.into_inner();
+        values.push(result);
+    }
+
+    // If there was only one retrieved value, there's no actual need to run the tally, as this means
+    // that only one transport was used and therefore the node is not in paranoid mode.
+    if values.len() > 1 {
+        let mut context = ReportContext::from_stage(Stage::Tally(TallyMetaData::default()));
+        let _tally = run_tally_with_context_report(values.clone(), tally, &mut context, settings);
+
+        // If at least one of the retrieved values is marked as an outlier, resolve to the
+        // `InconsistentSource` error.
+        if let Stage::Tally(TallyMetaData { liars, .. }) = context.stage {
+            if !liars.is_empty() {
```

Comment:
```suggestion
            if liars.iter().any(|&v| v==true) {
```

Reply:
Also, seems that as soon as there is some specific error (403), the paranoid_retrieval returns the error whitout evaluating the responses. This is, if locally I get a value  and through tor I get a 403, the 403 is returned and committed, I think we should refrain from commiting in that case